payroll/models/hr_salary_rule.py

```diff
@@ -178,77 +178,84 @@ def _recursive_search_of_rules(self):
-                raise UserError(
-                    _(
-                        "Wrong percentage base or quantity defined for salary "
-                        "rule %s (%s)."
-                    )
-                    % (self.name, self.code)
-                )
-        else:
-            try:
-                safe_eval(
-                    self.amount_python_compute, localdict, mode="exec", nocopy=True
-                )
-                result_qty = 1.0
-                result_rate = 100.0
-                result_name = False
-                if "result_name" in localdict:
-                    result_name = localdict["result_name"]
-                if "result_qty" in localdict:
-                    result_qty = localdict["result_qty"]
-                if "result_rate" in localdict:
-                    result_rate = localdict["result_rate"]
-                return (
-                    float(localdict["result"]),
-                    float(result_qty),
-                    float(result_rate),
-                    result_name,
+        method = "_compute_rule_{}".format(self.amount_select)
+        return api.call_kw(self, method, [self.ids, localdict], {})
+
+    def _compute_rule_fix(self, localdict):
+        try:
+            return {
+                "name": self.name,
+                "quantity": float(safe_eval(self.quantity, localdict)),
+                "rate": 100.0,
+                "amount": self.amount_fix,
+            }
+        except Exception:
+            raise UserError(
+                _("Wrong quantity defined for salary rule %s (%s).")
+                % (self.name, self.code)
+            )
+
+    def _compute_rule_percentage(self, localdict):
+        try:
+            return {
+                "name": self.name,
+                "quantity": float(safe_eval(self.quantity, localdict)),
+                "rate": self.amount_percentage,
+                "amount": float(safe_eval(self.amount_percentage_base, localdict)),
+            }
+        except Exception:
+            raise UserError(
+                _(
+                    "Wrong percentage base or quantity defined for salary "
+                    "rule %s (%s)."
                 )
-            except Exception as ex:
-                raise UserError(
-                    _(
-                        """
+                % (self.name, self.code)
+            )
+
+    def _compute_rule_code(self, localdict):
+        try:
+            safe_eval(self.amount_python_compute, localdict, mode="exec", nocopy=True)
+            return self._get_rule_dict(localdict)
+        except Exception as ex:
+            raise UserError(
+                _(
+                    """
 Wrong python code defined for salary rule %s (%s).
 Here is the error received:
 
 %s
 """
-                    )
-                    % (self.name, self.code, repr(ex))
                 )
+                % (self.name, self.code, repr(ex))
+            )
+
+    def _get_rule_dict(self, localdict):
```

Comment:
Why is the use of this method? Should this logic be inside `_compute_rule_code`? Is there a reason to have this in a diferent method?

Reply:
Yes, I split the method so that one rule may return values for multiple payslip lines.

* Run `safe_eval()` once, in `_compute_rule_code()`
* I will override `_get_rule_dict()` and call `super()` for each line.